lesson20/index.js

```diff
@@ -0,0 +1,130 @@
+const parent = document.querySelector("body");
+
+const createElementWithClassName = (type, className) => {
+    const element = document.createElement(type);
+    element.className = className;
+    return element;
+};
+
+const addLoading = () => {
+    const img = document.createElement("img");
+    const imgWrapper = createElementWithClassName("div", "loading");
+
+    img.src = "loading-circle.gif";
+    imgWrapper.id = "js-loading";
+    parent.appendChild(imgWrapper).appendChild(img);
+};
+
+const removeLoading = () => {
+    document.getElementById("js-loading").remove();
+};
+
+const fetchData = async(endpoint) => {
+    const response = await fetch(endpoint);
+
+    if(!response.ok){
+        const errorMessage = `${response.status}:${response.statusText}`;
+        parent.appendChild(createErrorMessage(errorMessage));
+        console.error(errorMessage);
+        return;
+    }
+    return await response.json();
+};
+
+const fetchUserData = async() => {
+    return new Promise((resolve) => {
+        setTimeout(() => resolve(fetchData("https://mocki.io/v1/8853f6f1-f1f0-4db3-9788-668933e2f788")), 3000);
+        // setTimeout(() => resolve(fetchData("https://mocki.io/v1/f12ae2d1-310d-4120-8749-47773d65e236")), 3000);//空配列
+        // setTimeout(() => resolve(fetchData("https://httpstat.us/503")), 3000);//503 error
+    })
+};
+
+const init = async() => {
+    addLoading();
+    try {
+        const json = await fetchUserData();
+        if (!json) return;
+
+        const data = json.data;
+        renderTable(data);
+
+        if (data.length === 0) {
+            parent.textContent = "まだデータがありません";
+            console.log("まだデータがありません");
+        }
```

Comment:
IMO
An error has occurred when the array is empty.
It is recommended to create an element to display the message and then add it to parent.
```suggestion
        if (data.length === 0) {
            const message = document.createElement("p");
            message.textContent = "まだデータがありません";
            parent.appendChild(message);
            console.log("まだデータがありません");
            return;
        }
        renderTable(data);
```

Reply:
I'm sorry, as Chihiro-san pointed out, I think it's a good idea to separate renderTable ().

```
const getData = async() => {
    addLoading();
    try {
        const json = await fetchUserData();
        if (!json) return;
        const data = json.data;

        if (data.length === 0) {
            const message = document.createElement("p");
            message.textContent = "まだデータがありません";
            parent.appendChild(message);
            console.log("まだデータがありません");
        }
        return data;

    } catch(e) {
        console.error(e);
        parent.appendChild(createErrorMessage(e));
    } finally {
        removeLoading();
    }
};

const init = async () => {
  const data = await getData();
  data && renderTable(data);
};
```

Reply:
Thank you for your review!!

Thanks for pointing out that 'return data' wasn't used, and that I should have used <p> to display the error message!
I hadn't noticed that, so thanks.

This time, I added <p> and the init() function following Chihiro's advice.

I learned a great deal! Thank you both.
[34ebdb3](https://github.com/haru-programming/JavaScript-lessons/pull/25/commits/34ebdb36bdd11d1e5163003c8b18798f5332358f)